File: LeetCode/Python/234.palindrome-linked-list.py
# @before-stub-for-debug-begin
from python3problem234 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=234 lang=python3
#
# [234] Palindrome Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# TODO: Do in O(1) Space and O(n) time
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        nodeLst = []

        tempHead = head
        while tempHead:
            nodeLst.append(tempHead.val)
            tempHead = tempHead.next


        lstLen = len(nodeLst)
        for i in range(lstLen // 2):
            if nodeLst[i] != nodeLst[lstLen - i - 1]:
                return False
        return True

    
# @lc code=end

File: LeetCode/CPP/234.palindrome-linked-list.cpp
/*
 * @lc app=leetcode id=234 lang=cpp
 *
 * [234] Palindrome Linked List
 */
#include <bits/stdc++.h>
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};
using namespace std;
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    bool isPalindrome(ListNode *head)
    {
        stack<int> s;
        ListNode *fast = head, *slow = head;

        while (fast && fast->next)
        {
            fast = fast->next->next;
            s.push(slow->val);
            slow = slow->next;
        }
        if (fast != nullptr)
            slow = slow->next;

        while (!s.empty() && slow != nullptr && s.top() == slow->val)
        {
            slow = slow->next;
            s.pop();
        }
        return s.empty();
    }
};
// @lc code=end
